internal/controllers/machineset/machineset_controller.go

```diff
@@ -355,6 +330,77 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 	return result, nil
 }
 
+func (r *Reconciler) reconcileDelete(ctx context.Context, machineSet *clusterv1.MachineSet) (reconcile.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	machineList, err := r.getMachinesForMachineSet(ctx, machineSet)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// If all the descendant machines are deleted, then remove the machinedeployment's finalizer.
+	if len(machineList) == 0 {
+		controllerutil.RemoveFinalizer(machineSet, clusterv1.MachineSetFinalizer)
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("MachineSet still has owned Machines, deleting them first")
+	for _, machine := range machineList {
+		if machine.DeletionTimestamp.IsZero() {
+			log.Info("Deleting Machine", "Machine", klog.KObj(machine))
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return ctrl.Result{}, errors.Wrapf(err, "failed to delete Machine %s", klog.KObj(machine))
+			}
+		}
+	}
+
+	return ctrl.Result{RequeueAfter: deleteRequeueAfter}, nil
+}
+
+func (r *Reconciler) getMachinesForMachineSet(ctx context.Context, machineSet *clusterv1.MachineSet) ([]*clusterv1.Machine, error) {
+	log := ctrl.LoggerFrom(ctx)
+	selectorMap, err := metav1.LabelSelectorAsMap(&machineSet.Spec.Selector)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to convert MachineSet %q label selector to a map", machineSet.Name)
+	}
+
+	// Get all Machines linked to this MachineSet.
+	allMachines := &clusterv1.MachineList{}
+	err = r.Client.List(ctx,
+		allMachines,
+		client.InNamespace(machineSet.Namespace),
+		client.MatchingLabels(selectorMap),
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list machines")
+	}
+
+	// Filter out irrelevant machines (i.e. IsControlledBy something else) and claim orphaned machines.
+	// Machines in deleted state are deliberately not excluded https://github.com/kubernetes-sigs/cluster-api/pull/3434.
+	filteredMachines := make([]*clusterv1.Machine, 0, len(allMachines.Items))
+	for idx := range allMachines.Items {
+		machine := &allMachines.Items[idx]
+		log := log.WithValues("Machine", klog.KObj(machine))
+		if shouldExcludeMachine(machineSet, machine) {
+			continue
+		}
+
+		// Attempt to adopt machine if it meets previous conditions and it has no controller references.
+		if metav1.GetControllerOf(machine) == nil {
+			if err := r.adoptOrphan(ctx, machineSet, machine); err != nil {
+				log.Error(err, "Failed to adopt Machine")
+				r.recorder.Eventf(machineSet, corev1.EventTypeWarning, "FailedAdopt", "Failed to adopt Machine %q: %v", machine.Name, err)
+				continue
+			}
+			log.Info("Adopted Machine")
+			r.recorder.Eventf(machineSet, corev1.EventTypeNormal, "SuccessfulAdopt", "Adopted Machine %q", machine.Name)
+		}
+
+		filteredMachines = append(filteredMachines, machine)
+	}
```

Comment:
Should we do this outside? Otherwise we're not retrieving and adopting the machines at the same time

cc @sbueringer

Reply:
@vincepri Do you mean "Otherwise we **are** retrieving and adopting the machines at the same time"?

I'm not sure if adopting here as well is a bad thing (we also now do the same in reconcileDelete in the MD controller).

Basically it allows us to also fixup ownerRef chains during reconcileDelete, which maybe helps to cover some edge cases?

Reply:
Yes I meant we're doing both, the trigger for this comment was that `getMachinesForMachineSet` doesn't mention we might also adopt them

Reply:
@chrischdi Sounds good

Reply:
(in both controllers IIRC)